modules/bridge/src/main/ts/ephox/bridge/components/dialog/Tree.ts

```diff
@@ -0,0 +1,96 @@
+import { FieldSchema, StructureSchema } from '@ephox/boulder';
+import { Optional, Result } from '@ephox/katamari';
+
+import { ToolbarMenuButtonSpec, ToolbarMenuButton } from '../../api/Toolbar';
+import * as ComponentSchema from '../../core/ComponentSchema';
+import { MenuButtonSchema } from '../toolbar/ToolbarMenuButton';
+
+type Id = string;
+
+export interface TreeSpec {
+  type: 'tree';
+  items: TreeItemSpec[];
+  onLeafAction?: (id: Id) => void;
+}
+
+export interface Tree {
+  type: 'tree';
+  items: TreeItem[];
+  onLeafAction: Optional<(id: Id) => void>;
+}
+
+interface BaseTreeItemSpec {
+  title: string;
+  id: Id;
+  menu?: ToolbarMenuButtonSpec;
+}
+
+interface BaseTreeItem {
+  title: string;
+  id: string;
+  menu: Optional<ToolbarMenuButton>;
+}
+
+export interface DirectorySpec extends BaseTreeItemSpec {
+  type: 'directory';
+  children: TreeItemSpec[];
+}
+
+export interface Directory extends BaseTreeItem {
+  type: 'directory';
+  children: TreeItem[];
+}
+
+export interface LeafSpec extends BaseTreeItemSpec {
+  type: 'leaf';
+}
+
+export interface Leaf extends BaseTreeItem {
+  type: 'leaf';
+}
+
+export type TreeItemSpec = DirectorySpec | LeafSpec;
+
+export type TreeItem = Directory | Leaf;
+
+const baseTreeItemFields = [
+  FieldSchema.requiredStringEnum('type', [ 'directory', 'leaf' ]),
+  ComponentSchema.title,
+  FieldSchema.requiredString('id'),
+  FieldSchema.optionOf('menu', MenuButtonSchema ),
+];
+
+const treeItemLeafFields = baseTreeItemFields;
+
+const treeItemLeafSchema = StructureSchema.objOf(treeItemLeafFields);
+
+const treeItemDirectoryFields = baseTreeItemFields.concat([
+  FieldSchema.requiredArrayOf('children', StructureSchema.thunkOf('children', () => {
+    // The order here is very important! if we change the order, then boulder will always match
+    // againt treeItemLeafSchema, because a directory has all attributes of leaf plus the children property
```

Comment:
```suggestion
    // against treeItemLeafSchema, because a directory has all attributes of leaf plus the children property
```

I thought it was picking a validator based on the `type`?

Reply:
Yeah this is an outdated comment. I removed it now. Good catch